### PR TITLE
feat(060): forum category vocabulary, guarded retirement mutation + category-change validation

### PR DESCRIPTION
**Pull request by bobbykolev**
## What & why

Consolidates the Alkemio Forum toward four clearer categories — Releases, Newsletter, Tips & Tricks, Q&A — per [alkem-io/alkemio#2052](https://github.com/alkem-io/alkemio/issues/2052). This is the **server** slice (wave 1, sole contract producer).

Adds two vocabulary members, propagates them to the platform forum's active list by migration, closes a data-integrity gap on the category-change path, and ships a guarded retirement mutation so the four obsolete categories can be retired **after** the human curation pass without a further code release.

Workspace feature: `workspace#060-forum-reorganisation`

## Changes

- **Two new vocabulary members** — `NEWSLETTER` (`newsletter`), `TIPS_AND_TRICKS` (`tips-and-tricks`). No member is ever removed: members are permanent tombstones, because deleting one while any row still carries its value fails GraphQL enum serialization and nulls the *entire* platform query.
- **Append-only idempotent migration** appends both values to the single Forum row's active list, shipped in the same release as the code. `ensureForumCreated` keeps its create-only-when-absent behaviour — no boot-time reconciliation writes.
- **Rollback-safe read filters.** Unknown stored values are dropped from `Forum.discussionCategories`, and `Discussion.category` is guarded symmetrically, so an older or rolled-back build meeting a value it does not know degrades to "not offered" instead of nulling the forum page or the discussions list.
- **Shared `assertForumCategoryAllowed`** applied to **both** create and category-change paths. Previously `updateDiscussion` assigned `category` straight from input with no validation, so "this category is empty" could never be held as an invariant.
- **Data-driven `ADMIN_ONLY_FORUM_CATEGORIES`** (`{RELEASES, NEWSLETTER}`) replaces the hardcoded `category === RELEASES` comparison, enforced on both paths.
- **New `adminForumRemoveDiscussionCategory` mutation** — `PLATFORM_ADMIN`-gated, audited, idempotent, removal-only. Refuses with a typed `FORUM_DISCUSSION_CATEGORY_NOT_EMPTY` error **naming the live post count** while any post still carries the category, so "once empty" is a server-enforced precondition rather than a claim in a PR description. The vocabulary member is never touched.
- **Read-only boot drift log** (never a write) makes an unrun migration observable.

## Design decisions worth reviewing

- **Q&A is the existing Help category relabelled in place** — value, slug, `/forum/help` URL and Matrix space key all stay `help`; only the display label changes. A UI reading "Q&A" over a `/forum/help` URL is intentional, not a bug. Every existing Help post therefore becomes a Q&A post without individual review — an accepted, recorded consequence.
- **No add-category mutation, by design.** Adding a category inherently needs code (enum member, URL slug, icon, six locales), so it rides a release; retirement needs only the allow-list edit, which is why it gets a guarded mutation.
- **An authorization denial writes no audit record.** The gate is evaluated *before* the audited block, matching every sibling audited platform-operations mutation. Post-authorization outcomes — success, idempotent no-op, and the not-empty refusal — are all still audited. See the security note below.
- **No Matrix re-parenting on category change.** matrix-adapter exposes no removal primitive (`GetSpaceChildren`/`AddSpaceChild`/`SetSpaceParent` only), so "moving" a room yields permanent, unrepairable dual parentage. Stale parentage is deliberate; please do not "fix" it.
- **Recategorisation sends no notification** — a curation sitting over N posts must not become N platform-wide sends.

## Security

A security review found and this PR fixes a defect introduced during implementation: `grantAccessOrFail` had been placed *inside* the audited try/catch, so every rejected call attempted a write to `platform_audit_entry` and emitted an ERROR log. Because `AuthInterceptor` resolves a credential-less request to an anonymous actor rather than returning 401, an **unauthenticated caller** could drive an unbounded stream of failed INSERTs and error logs; any registered user, having a valid UUID, could make those INSERTs **succeed** and append rows to the compliance table. Fixed by hoisting authorization above the audited block.

Verified by A/B on the live audit table: non-admin and anonymous denials → **11 → 11 rows** (none written); admin hitting the not-empty refusal → **11 → 12** (exactly one, as required).

SOC 2 CC7.2/CC7.3/CC4.1/A1.1 · ISO 27001 A.8.15/A.8.16/A.8.6.

## Verification

- Gates green: `pnpm lint`, `pnpm test:ci:no:coverage` (**8954 tests**), `pnpm build`.
- Schema diff: **0 BREAKING**, 4 additive.
- Live stack walk of 24 acceptance scenarios, including the full retirement behaviour table (typed not-empty refusal naming the count, removal, idempotent retry, non-admin refusal, audit rows) and the tombstone guarantee (a post carrying a *retired* category still serialises, renders and can be moved out).
- Adversarial review: 106 findings raised, 21 killed by verification, 8 confirmed, all fixed.

## Known gaps (tracked, not silently dropped)

- The `Forum.discussionCategories` read filter is covered by unit tests but was **never exercised live** — it is the mitigation for the highest-rated risk here.
- Historical in-app notification label rendering was unobservable (empty notification table on the verification stack).

## Rollout

Run the migration with the release. Afterwards, run the existing audited `adminCommunicationSyncSpaceHierarchy` mutation to create the Matrix spaces for the two new categories (they are otherwise created lazily on first post).

Spec, plan and evidence ledger: `agents-hq/specs/060-forum-reorganisation/`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **New Features**
  - Added **Newsletter** and **Tips & Tricks** discussion categories.
  - Platform administrators can remove active forum categories when no discussions use them.
  - Added auditing for forum category changes.

- **Improvements**
  - Newsletter discussions now require platform-admin authorization.
  - Unknown categories are safely handled and excluded from displayed categories.
  - Category removal is blocked when discussions still use the category.

- **Maintenance**
  - Added safeguards to detect differences between configured and stored forum categories.

### CODE REVIEW

File: src/platform/forum/forum.service.ts [hunk 1]
```diff
@@ -310,6 +309,55 @@ export class ForumService {
...
 
+  /**
+   * Loads the platform's singleton Forum row directly, bypassing
+   * `PlatformService` (which itself depends on `ForumService` — injecting
+   * it here would create a module cycle). The platform bootstrap always
+   * seeds exactly one Forum, so absence here means the platform itself was
+   * never bootstrapped.
+   */
+  async getPlatformForumOrFail(): Promise<IForum> {
+    const forum = await this.forumRepository.findOne({ where: {} });
```
[the remark this answers is not in the document]
Comment (bobbykolev) [reply src/platform/forum/forum.service.ts]: **Addressed** in `df0b8313a`.

Confirmed on the merits. `getPlatformForumOrFail()` used `forumRepository.findOne({ where: {} })` with no ordering, so *any* `forum` row was an equally valid candidate — a partially rolled-back bootstrap, a restored dump, or a future non-platform Forum. `adminForumRemoveDiscussionCategory` is the one write path that consumes it, so the failure mode is shrinking the wrong row's active category list. "There is only one row today" is an assumption about data, not an invariant the code holds.

Fix traverses the relation, which is what actually makes a Forum *the platform's*:

```ts
const platform = await this.platformRepository.findOne({
  where: {},
  relations: { forum: true },
});
if (!platform?.forum) throw new EntityNotFoundException(...);
return platform.forum;
```

Resolved through the `Platform` **repository** rather than `PlatformService.getForumOrFail()` — that service depends on `ForumService`, so injecting it back here would close a module cycle. `ForumModule` now registers `TypeOrmModule.forFeature([Forum, Platform])`.

Specs assert the relation is used, that the `forum` table is never scanned (`expect(repo.findOne).not.toHaveBeenCalled()`), and both absence paths (no platform row / platform carrying no forum).

File: src/platform/forum-discussion/discussion.resolver.mutations.ts [hunk 2]
```diff
@@ -55,9 +65,63 @@ export class DiscussionResolverMutations {
...
+    // the create path requires for these same
+    // categories — this makes "the active list defines what's allowed"
+    // an invariant the server enforces on category-change too, not only
+    // on create.
+    if (updateData.category) {
+      if (!discussion.forum) {
+        throw new EntityNotFoundException(
+          `Unable to load Forum for Discussion with ID: ${discussion.id}`,
+          LogContext.PLATFORM_FORUM
+        );
```
[the remark this answers is not in the document]
Comment (bobbykolev) [reply src/platform/forum-discussion/discussion.resolver.mutations.ts:96]: **Addressed** in `c5e5e913c`.

The *disclosure* premise is weak here — `getDiscussionOrFail` matches on `id` only, so `discussion.id === updateData.ID`, i.e. the message echoes the caller's own input back. Nothing is disclosed that the caller did not just send.

Taking it anyway, on two stronger grounds:

1. `CLAUDE.md` → *Exception Handling* states the repo convention outright: dynamic data goes in the structured `details` property, and `EntityNotFoundException` already accepts it as its third argument.
2. Message quality. This branch does not fail because an id was not found — it fails because the `forum` **relation** did not load. The id is operator diagnostics, not user-actionable, so it belongs in `details` (which the filter still logs server-side) rather than in the text the user reads.

```ts
throw new EntityNotFoundException(
  'Unable to load Forum for Discussion',
  LogContext.PLATFORM_FORUM,
  { discussionID: discussion.id }
);
```

Note that the surrounding guard also moved from `if (updateData.category)` to a real change comparison in the same commit, so this branch is now reachable only on an actual recategorisation.

File: src/common/exceptions/forum.discussion.category.not.empty.exception.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Thrown by `adminForumRemoveDiscussionCategory` when the requested category
+ * still holds at least one Discussion. Deliberately a distinct code from
+ * `ForumDiscussionCategoryException` so ops tooling can tell "category not
+ * allowed" apart from "category not empty".
+ */
+export class ForumDiscussionCategoryNotEmptyException extends BaseException {
+  constructor(category: ForumDiscussionCategory, remainingPostCount: number) {
+    super(
+      `Cannot remove discussion category '${category}': ${remainingPostCount} post(s) still carry it`,
```
[the remark this answers is not in the document]
Comment (bobbykolev) [reply src/common/exceptions/forum.discussion.category.not.empty.exception.ts:14]: **Declined** — for this exception the dynamic content is the requirement, and `details` cannot carry it where it is needed.

The convention you are invoking is real (`CLAUDE.md` → *Exception Handling*), and we applied it in the sibling thread on `discussion.resolver.mutations.ts:82`. It does not reach here, for three independent reasons:

**1. The count is a specified part of the error, not incidental.** The feature spec requires the not-empty refusal to state the remaining post count (FR-012(i), US3-AS1, SC-003: *"refused 100% of the time while the category holds ≥ 1 post, with an error a human can act on (names the count)"*). It is the whole point of the distinct error code: ops tooling distinguishes *not allowed* from *not empty*, and the operator needs to know how many posts are left to curate.

**2. `details` is stripped in production — the exact environment the requirement exists for.** `GraphqlExceptionFilter` returns a *new* `GraphQLError` carrying only `errorId` and `code` when `NODE_ENV === 'production'`; `details`, `numericCode` and `userMessage` are all dropped. Moving the count into `details` would therefore not relocate it — it would delete it from every production response, silently, while looking like a tidy-up.

**3. A live integration test asserts on the message text.** `test-suites` → `platform-discussions.it-spec.ts`, *"Admin cannot remove a category that still holds a post; the count is in the error…"* regex-matches the count out of `errors[0].message`. That test runs against real environments where `details` is unavailable, which is precisely why it reads the message.

On the values themselves: neither is an identifier or PII. `remainingPostCount` is an aggregate; `category` is a public `ForumDiscussionCategory` enum member the caller supplied in `removeData.category` and which is published in `schema.graphql`. The convention's own examples are *IDs, emails*. Both values are already in `details` as well, for the non-production/log path.

**On the two `grantAccessOrFail` `msg` sites** (`forum.resolver.mutations.ts`, `discussion.resolver.mutations.ts`): also declined. `grantAccessOrFail` builds its own `errorMsg` as ``Authorization: unable to grant '<privilege>' privilege: <msg> user: <actorID> on authorization <auth.id> of type '<type>'`` — the actor UUID and the policy UUID are appended by the *framework*, at every one of the hundreds of call sites in `src/`, including the two pre-existing lines in this same file (`delete discussion: ${discussion.id}`, `Update discussion: ${discussion.id}`). Making our two `msg` arguments static would remove a public enum name while leaving two genuine UUIDs in place — no security benefit, and inconsistent with the surrounding code. If the framework-appended identifiers are worth addressing, that is a repo-wide change to `AuthorizationService`, not something to start unevenly in this PR.

File: src/migrations/1788300000000-AddForumCategoriesNewsletterTipsTricks.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        category => !existing.includes(category)
+      );
+      if (missing.length === 0) {
+        continue;
+      }
+
+      const updated = [...existing, ...missing].join(',');
+      await queryRunner.query(
+        `UPDATE forum SET "discussionCategories" = $1 WHERE id = $2`,
+        [updated, row.id]
```
[the remark this answers is not in the document]
Comment (bobbykolev) [reply src/migrations/1788300000000-AddForumCategoriesNewsletterTipsTricks.ts:130]: **Addressed** in `9610409f8` — took the row lock; declined the "stop the writers" half.

The race is real and reachable, and slightly more so than the finding states. The retirement mutation ships in *this same release*, and the documented rollout is deliberately **code-first, then migrate** (the migration job is manually triggered). So there is a genuine window in which the mutation is live in the running server and this migration has not run yet — a retirement committing between the `SELECT` and the full-list `UPDATE` would be overwritten and the category resurrected. The derivation is already retirement-safe for anything committed *before* the read (it filters the canonical order down to what is present, and only ever adds the two named values); the lock closes the remaining gap.

```sql
SELECT id, "discussionCategories" FROM forum FOR UPDATE
```

in both `up()` and `down()`. Cost/benefit is one-sided: one row, in a single-row table, held for the length of a migration transaction that does one derivation and one write. A concurrent retirement simply waits rather than being silently lost. No optimistic-concurrency column and no operational "stop the writers" runbook step are warranted for that — those would add coordination burden for a mutation used roughly four times, ever.

Specs assert the lock is taken on both paths (`selects()[0]` contains `FOR UPDATE`, exactly one select per run).

File: src/platform/forum/forum.service.ts [hunk 5]
```diff
@@ -310,6 +309,55 @@ export class ForumService {
...
+  async removeDiscussionCategory(
+    forum: IForum,
+    category: ForumDiscussionCategory
+  ): Promise<{ forum: IForum; removed: boolean }> {
+    if (!forum.discussionCategories.includes(category)) {
+      return { forum, removed: false };
+    }
+
+    const remainingPostCount =
+      await this.discussionService.countInForumByCategory(forum.id, category);
```
[the remark this answers is not in the document]
Comment (bobbykolev) [reply src/platform/forum/forum.service.ts:358]: **Declined** — this is the micro-race recorded and accepted as **A-09** in the feature spec (`specs/060-forum-reorganisation/spec.md`, Clarifications → adopted council agreements).

A-09, verbatim: *"a post created into a category concurrently with its removal (between count and write) lands in a just-retired category: readable at permalink and category URL (tombstone), recategorisable out; no transactional ceremony for an admin operation used ~4 times."* It is also written up as a named edge case and as part of the US3 independent test.

I assessed whether your framing is genuinely worse than what A-09 records. It is not — it is the same race described from the other side. A-09 measures the window from the removal path (count → list update); you measure it from the writer path (validate → `DiscussionService.save`). Both describe the same interleaving and produce the same single outcome: **one post carrying a category that is no longer active.** The consequence set is identical, and the spec already enumerates it:

- The post stays readable at its permalink and at the retired category's URL — vocabulary members are permanent tombstones (A-02, FR-002), and the read-side drift guard resolves any unrecognised stored category to the `Other` tombstone rather than failing enum serialization (FR-007/FR-007a). Nothing 500s, nothing is orphaned.
- The post can be recategorised out: the edit picker deliberately offers the active list **plus** the post's own current category precisely so a post in a retired category can be moved out (FR-009 / US2-AS3). That path is now also fixed for the echo-the-same-category case in `c5e5e913c`.
- The mutation is idempotent (US3-AS3, FR-012(iv)), so the operator's recovery is exactly the normal loop: re-run it, get the typed not-empty refusal naming the count, move the post, re-run. No ops intervention, no data repair.

The remedy you propose — a shared `Forum`-row lock spanning *creation, recategorisation and removal*, with revalidation immediately before every `DiscussionService.save` — puts a lock on the platform Forum row into the hot path of **every forum post write**, permanently, to protect an administrative operation that will be invoked about four times in the product's life, during a post-curation window, by one platform admin. That is the transactional ceremony A-09 explicitly weighed and declined. Note this is also materially different from the migration lock in the sibling thread, which I did take: that one is confined to a one-off maintenance transaction and touches no user-facing write path.

Re-raising a recorded, reasoned acceptance is out of scope for this PR. If the trade-off is to be revisited, the place is the spec's decision log, not a review thread.

File: src/migrations/1788300000000-AddForumCategoriesNewsletterTipsTricks.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    'platform-functionalities',
+    'community-building',
+    'challenge-centric',
+    'other',
+  ];
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    const rows: { id: string; discussionCategories: string | null }[] =
+      await queryRunner.query(
+        `SELECT id, "discussionCategories" FROM forum FOR UPDATE`
```
[the remark this answers is not in the document]
Comment (bobbykolev) [reply src/migrations/1788300000000-AddForumCategoriesNewsletterTipsTricks.ts]: Addressed in `4be024fc8`.

You were right, and this was the last instance of a defect already fixed elsewhere on this branch: `729bc60dc` changed `getPlatformForumOrFail()` off `findOne({ where: {} })` for exactly this reason. The migration was the remaining unscoped reader.

Both `up()` and `down()` now resolve the target through the `platform."forumId"` FK and lock only that row:

```sql
SELECT f.id, f."discussionCategories" FROM forum f
INNER JOIN platform p ON p."forumId" = f.id
FOR UPDATE OF f
```

`FOR UPDATE OF f` locks the `forum` row only, not the joined `platform` row, and the lock is still taken on the same SELECT that drives the write, so the ordering the row-lock commit established is preserved.

If no platform forum resolves (a database whose platform row has a NULL `forumId`, e.g. pre-bootstrap), the join returns zero rows and the migration is a clean no-op — it never throws and never falls back to scanning `forum`.

Added the migration test you asked for: two `forum` rows seeded with only one wired to `platform."forumId"`, asserting the platform forum is reordered/extended while the other row stays byte-identical, plus the `down()` equivalent (a non-platform forum carrying both added values is left untouched while the platform forum has them stripped) and two no-platform-forum no-op cases. The tests were verified against the pre-fix migration and 4 of 16 failed as expected, so they genuinely catch this regression rather than merely passing.

Idempotency, retirement-safety (never resurrecting a category absent from the column) and unknown-value preservation are unchanged and still covered. `schema.graphql` did not move.

File: src/migrations/1788300000000-AddForumCategoriesNewsletterTipsTricks.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+      await queryRunner.query(
+        `UPDATE forum SET "discussionCategories" = $1 WHERE id = $2`,
+        [updated, row.id]
+      );
+      console.log(
+        `[Migration] AddForumCategoriesNewsletterTipsTricks: reordered forum ${row.id} to ${JSON.stringify(
+          updated
+        )}`
+      );
```
[the remark this answers is not in the document]
Comment (bobbykolev) [reply src/migrations/1788300000000-AddForumCategoriesNewsletterTipsTricks.ts]: Declining this one, on the repo's own evidence.

`console.log` is the established convention for migrations here, and a Winston logger is not reachable in that context. Counted across `src/migrations` (117 migrations):

- **16** use `console.log`
- **0** use `WINSTON_MODULE_NEST_PROVIDER`, `LoggerService`, or `createLogger`

That split isn't accidental. Migrations execute through the TypeORM CLI (`ts-node`), outside the Nest DI container, so there is no injected `LoggerService` to reach — the logger contract you're pointing at is constructed by the Nest application factory, which never runs in a migration process. Adopting it here would mean either standing up a bespoke Winston instance inside the migration or reaching for a global, and would make this file the only one of 117 doing so.

Existing precedent, e.g. `1764590884533-seed.ts`:

```ts
console.log('Seeding initial platform data...');
```

Happy to revisit if the repo later adds a migration-safe logging helper — at that point it should be applied across all 16 migrations in one change rather than introduced unilaterally in this one.

The two message strings were updated in the scoping commit to say "platform forum", since the migration now targets that row specifically.

**Review by Comoque1 — APPROVED** [review]

**Review by antst — CHANGES_REQUESTED** [review]
Sync with matrix side is missing

**Comment by bobbykolev** [comment]
> Sync with matrix side is missing @antst 

Matrix part implemented. I need this on merged in order to continue merging the next ones:

https://github.com/alkem-io/server/pull/6484
https://github.com/alkem-io/matrix-adapter/pull/71
